esi/src/lib.rs

```diff
@@ -23,210 +22,395 @@
+                            debug!("nothing waiting so streaming directly to client");
+                            output_writer.write_event(event)?;
+                            output_writer
+                                .inner()
+                                .flush()
+                                .expect("failed to flush output");
+                        } else {
+                            debug!("pushing content to buffer, len: {}", elements.len());
+                            let mut vec = Vec::new();
+                            let mut writer = Writer::new(&mut vec);
+                            writer.write_event(event)?;
+                            elements.push_back(Element::Raw(vec));
+                        }
                     }
                 }
+
+                poll_elements(
+                    &mut elements,
+                    output_writer,
+                    dispatch_fragment_request,
+                    process_fragment_response,
+                )?;
+
                 Ok(())
             },
         )?;
 
+        // Wait for any pending requests to complete
+        loop {
+            if elements.is_empty() {
+                break;
+            }
+
+            poll_elements(
+                &mut elements,
+                output_writer,
+                dispatch_fragment_request,
+                process_fragment_response,
+            )?;
+        }
+
         Ok(())
     }
+}
+
+fn build_fragment_request(mut request: Request, url: &str) -> Result<Request> {
+    let escaped_url = match quick_xml::escape::unescape(url) {
+        Ok(url) => url,
+        Err(err) => {
+            return Err(ExecutionError::InvalidRequestUrl(err.to_string()));
+        }
+    }
+    .to_string();
 
-    fn send_esi_fragment_request(
-        &self,
-        original_request: &Request,
-        url: &str,
-        request_handler: &dyn Fn(Request) -> Result<Response>,
-    ) -> Result<Response> {
-        let mut req = original_request.clone_without_body();
+    if escaped_url.starts_with('/') {
+        match Url::parse(
+            format!("{}://0.0.0.0{}", request.get_url().scheme(), escaped_url).as_str(),
+        ) {
+            Ok(u) => {
+                request.get_url_mut().set_path(u.path());
+                request.get_url_mut().set_query(u.query());
+            }
+            Err(_err) => {
+                return Err(ExecutionError::InvalidRequestUrl(escaped_url));
+            }
+        };
+    } else {
+        request.set_url(match Url::parse(&escaped_url) {
+            Ok(url) => url,
+            Err(_err) => {
+                return Err(ExecutionError::InvalidRequestUrl(escaped_url));
+            }
+        });
+    }
 
-        if url.starts_with('/') {
-            req.get_url_mut().set_path(url);
-        } else {
-            req.set_url(url);
+    let hostname = request.get_url().host().expect("no host").to_string();
+
+    request.set_header(header::HOST, &hostname);
```

Comment:
I'm no expert on C@E behavior. Is it necessary to explicitly set the `Host` header from the request URL? I'd have expected that would happen automatically. I know it's often overridden anyway by whatever was configured for the backend.